// @antv/g-plugin-svg-picker/dist/index.esm.js
import { __awaiter, __generator, __values, __extends } from 'tslib';
import { AbstractRendererPlugin } from '@antv/g-lite';

/**
 * pick shape(s) with Mouse/Touch event
 *
 * 1. find AABB with r-tree
 * 2. use elementFromPoint
 */
var SVGPickerPlugin = /** @class */ (function () {
    function SVGPickerPlugin() {
    }
    SVGPickerPlugin.prototype.apply = function (context, runtime) {
        var _this = this;
        var doc = context.config.document, renderingService = context.renderingService, 
        // @ts-ignore
        svgElementMap = context.svgElementMap;
        renderingService.hooks.pick.tapPromise(SVGPickerPlugin.tag, function (result) { return __awaiter(_this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.pick(svgElementMap, doc, result)];
            });
        }); });
        renderingService.hooks.pickSync.tap(SVGPickerPlugin.tag, function (result) {
            return _this.pick(svgElementMap, doc, result);
        });
    };
    SVGPickerPlugin.prototype.pick = function (svgElementMap, doc, result) {
        var _a, e_1, _b;
        var topmost = result.topmost, clientX = (_a = result.position, _a.clientX), clientY = _a.clientY;
        try {
            var targets = [];
            try {
                // @see https://developer.mozilla.org/zh-CN/docs/Web/API/Document/elementsFromPoint
                for (var _c = __values((doc || document).elementsFromPoint(clientX, clientY)), _d = _c.next(); !_d.done; _d = _c.next()) {
                    var element = _d.value;
                    var target = svgElementMap.get(element);
                    // don't need to account for `visibility` since DOM API already does
                    if (target && target.isInteractive()) {
                        targets.push(target);
                        if (topmost) {
                            result.picked = targets;
                            return result;
                        }
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_d && !_d.done && (_b = _c.return)) _b.call(_c);
                }
                finally { if (e_1) throw e_1.error; }
            }
            result.picked = targets;
        }
        catch (e) {
            result.picked = [];
        }
        return result;
    };
    SVGPickerPlugin.tag = 'SVGPicker';
    return SVGPickerPlugin;
}());

var Plugin = /** @class */ (function (_super) {
    __extends(Plugin, _super);
    function Plugin() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.name = 'svg-picker';
        return _this;
    }
    Plugin.prototype.init = function () {
        this.addRenderingPlugin(new SVGPickerPlugin());
    };
    Plugin.prototype.destroy = function () {
        this.removeAllRenderingPlugins();
    };
    return Plugin;
}(AbstractRendererPlugin));

export { Plugin };
