// @antv/f-engine/es/component/index.js
var Component = /** @class */function () {
  function Component(props, context, updater) {
    this.isMounted = false;
    // State 内部私有属性
    this.destroyed = false;
    this.props = props;
    this.state = {};
    this.context = context;
    this.updater = updater;
  }
  Component.prototype.willMount = function () {};
  Component.prototype.didMount = function () {};
  Component.prototype.shouldUpdate = function (_nextProps) {
    return true;
  };
  Component.prototype.willReceiveProps = function (_props, _context) {};
  Component.prototype.willUpdate = function () {};
  Component.prototype.didUpdate = function () {};
  Component.prototype.render = function () {
    return null;
  };
  Component.prototype.willUnmount = function () {};
  Component.prototype.didUnmount = function () {};
  Component.prototype.setState = function (partialState, callback) {
    if (this.destroyed) {
      return;
    }
    this.updater.enqueueSetState(this, partialState, callback);
  };
  Component.prototype.forceUpdate = function (callback) {
    if (this.destroyed) {
      return;
    }
    this.updater.enqueueForceUpdate(this, {}, callback);
  };
  Component.prototype.setAnimate = function (animate) {
    this.animate = animate;
    this._vNode.animate = animate;
  };
  Component.prototype.destroy = function () {
    this.destroyed = true;
    this.animator = null;
  };
  return Component;
}();
// 标识是否是组件
// @ts-ignore
Component.prototype.isF2Component = true;
export default Component;
