// @antv/g-mobile-webgl/dist/index.js
'use strict';

var tslib = require('tslib');
var gLite = require('@antv/g-lite');
var DeviceRenderer = require('@antv/g-plugin-device-renderer');
var DragDropEvent = require('@antv/g-plugin-dragndrop');
var HTMLRenderer = require('@antv/g-plugin-html-renderer');
var ImageLoader = require('@antv/g-plugin-image-loader');
var DomInteraction = require('@antv/g-plugin-mobile-interaction');
var WebGLDevice = require('@antv/g-plugin-webgl-device');
var util = require('@antv/util');

function _interopNamespaceDefault(e) {
    var n = Object.create(null);
    if (e) {
        Object.keys(e).forEach(function (k) {
            if (k !== 'default') {
                var d = Object.getOwnPropertyDescriptor(e, k);
                Object.defineProperty(n, k, d.get ? d : {
                    enumerable: true,
                    get: function () { return e[k]; }
                });
            }
        });
    }
    n.default = e;
    return Object.freeze(n);
}

var DeviceRenderer__namespace = /*#__PURE__*/_interopNamespaceDefault(DeviceRenderer);
var DragDropEvent__namespace = /*#__PURE__*/_interopNamespaceDefault(DragDropEvent);
var HTMLRenderer__namespace = /*#__PURE__*/_interopNamespaceDefault(HTMLRenderer);
var ImageLoader__namespace = /*#__PURE__*/_interopNamespaceDefault(ImageLoader);
var DomInteraction__namespace = /*#__PURE__*/_interopNamespaceDefault(DomInteraction);
var WebGLDevice__namespace = /*#__PURE__*/_interopNamespaceDefault(WebGLDevice);

function isCanvasElement(el) {
    if (!el || typeof el !== 'object')
        return false;
    if (el.nodeType === 1 && el.nodeName) {
        // HTMLCanvasElement
        return true;
    }
    // CanvasElement
    return !!el.isCanvasElement;
}

var WebGLContextService = /** @class */ (function () {
    function WebGLContextService(context) {
        this.canvasConfig = context.config;
        // @ts-ignore
        this.deviceRendererPlugin = context.deviceRendererPlugin;
    }
    WebGLContextService.prototype.init = function () {
        return tslib.__awaiter(this, void 0, void 0, function () {
            var _a, canvas, devicePixelRatio, dpr;
            return tslib.__generator(this, function (_b) {
                _a = this.canvasConfig, canvas = _a.canvas, devicePixelRatio = _a.devicePixelRatio;
                this.$canvas = canvas;
                // 实际获取到小程序环境的上下文
                this.context = this.$canvas.getContext('webgl');
                dpr = devicePixelRatio || 1;
                dpr = dpr >= 1 ? Math.ceil(dpr) : 1;
                this.dpr = dpr;
                this.resize(this.canvasConfig.width, this.canvasConfig.height);
                return [2 /*return*/];
            });
        });
    };
    WebGLContextService.prototype.getContext = function () {
        return this.context;
    };
    WebGLContextService.prototype.getDomElement = function () {
        return this.$canvas;
    };
    WebGLContextService.prototype.getDPR = function () {
        return this.dpr;
    };
    WebGLContextService.prototype.getBoundingClientRect = function () {
        if (this.$canvas.getBoundingClientRect) {
            return this.$canvas.getBoundingClientRect();
        }
    };
    WebGLContextService.prototype.destroy = function () {
        // TODO: 小程序环境销毁 context
        this.context = null;
        this.$canvas = null;
    };
    WebGLContextService.prototype.resize = function (width, height) {
        var pixelRatio = devicePixelRatio;
        var canvasDOM = this.$canvas; // HTMLCanvasElement or canvasElement
        // 浏览器环境设置style样式
        // @ts-ignore 使用style功能判断不使用类型判断更准确
        if (canvasDOM.style) {
            // @ts-ignore
            canvasDOM.style.width = width + 'px';
            // @ts-ignore
            canvasDOM.style.height = height + 'px';
        }
        if (isCanvasElement(canvasDOM)) {
            canvasDOM.width = width * pixelRatio;
            canvasDOM.height = height * pixelRatio;
            // if (pixelRatio !== 1) {
            //   this.context.scale(pixelRatio, pixelRatio);
            // }
        }
    };
    WebGLContextService.prototype.applyCursorStyle = function (cursor) {
        // 小程序环境无需设置鼠标样式
    };
    WebGLContextService.prototype.toDataURL = function (options) {
        return tslib.__awaiter(this, void 0, void 0, function () {
            return tslib.__generator(this, function (_a) {
                return [2 /*return*/, this.deviceRendererPlugin.toDataURL(options)];
            });
        });
    };
    return WebGLContextService;
}());

var ContextRegisterPlugin = /** @class */ (function (_super) {
    tslib.__extends(ContextRegisterPlugin, _super);
    function ContextRegisterPlugin(rendererPlugin) {
        var _this = _super.call(this) || this;
        _this.rendererPlugin = rendererPlugin;
        _this.name = 'mobile-webgl-context-register';
        return _this;
    }
    ContextRegisterPlugin.prototype.init = function () {
        this.context.ContextService = WebGLContextService;
        // @ts-ignore
        this.context.deviceRendererPlugin = this.rendererPlugin;
    };
    ContextRegisterPlugin.prototype.destroy = function () {
        delete this.context.ContextService;
    };
    return ContextRegisterPlugin;
}(gLite.AbstractRendererPlugin));

var Renderer = /** @class */ (function (_super) {
    tslib.__extends(Renderer, _super);
    function Renderer(config) {
        var _this = _super.call(this, config) || this;
        var deviceRendererPlugin = new DeviceRenderer__namespace.Plugin();
        _this.registerPlugin(new ContextRegisterPlugin(deviceRendererPlugin));
        _this.registerPlugin(new ImageLoader__namespace.Plugin());
        _this.registerPlugin(new WebGLDevice__namespace.Plugin((config === null || config === void 0 ? void 0 : config.targets)
            ? {
                targets: config.targets,
            }
            : {
                targets: ['webgl2', 'webgl1'],
            }));
        _this.registerPlugin(deviceRendererPlugin);
        _this.registerPlugin(new DomInteraction__namespace.Plugin());
        _this.registerPlugin(new HTMLRenderer__namespace.Plugin());
        _this.registerPlugin(new DragDropEvent__namespace.Plugin({
            isDocumentDraggable: util.isNil(config === null || config === void 0 ? void 0 : config.isDocumentDraggable) ? true : config.isDocumentDraggable,
            isDocumentDroppable: util.isNil(config === null || config === void 0 ? void 0 : config.isDocumentDroppable) ? true : config.isDocumentDroppable,
            dragstartDistanceThreshold: util.isNil(config === null || config === void 0 ? void 0 : config.dragstartDistanceThreshold)
                ? 10
                : config.dragstartDistanceThreshold,
            dragstartTimeThreshold: util.isNil(config === null || config === void 0 ? void 0 : config.dragstartTimeThreshold)
                ? 50
                : config.dragstartTimeThreshold,
        }));
        return _this;
    }
    return Renderer;
}(gLite.AbstractRenderer));

exports.DeviceRenderer = DeviceRenderer__namespace;
exports.HTMLRenderer = HTMLRenderer__namespace;
exports.DomInteraction = DomInteraction__namespace;
exports.WebGLDevice = WebGLDevice__namespace;
exports.Renderer = Renderer;
