// @antv/g-mobile-webgl/dist/index.esm.js
import { __awaiter, __generator, __extends } from 'tslib';
import { AbstractRendererPlugin, AbstractRenderer } from '@antv/g-lite';
import * as DeviceRenderer from '@antv/g-plugin-device-renderer';
export { DeviceRenderer };
import * as DragDropEvent from '@antv/g-plugin-dragndrop';
import * as HTMLRenderer from '@antv/g-plugin-html-renderer';
export { HTMLRenderer };
import * as ImageLoader from '@antv/g-plugin-image-loader';
import * as DomInteraction from '@antv/g-plugin-mobile-interaction';
export { DomInteraction };
import * as WebGLDevice from '@antv/g-plugin-webgl-device';
export { WebGLDevice };
import { isNil } from '@antv/util';

function isCanvasElement(el) {
    if (!el || typeof el !== 'object')
        return false;
    if (el.nodeType === 1 && el.nodeName) {
        // HTMLCanvasElement
        return true;
    }
    // CanvasElement
    return !!el.isCanvasElement;
}

var WebGLContextService = /** @class */ (function () {
    function WebGLContextService(context) {
        this.canvasConfig = context.config;
        // @ts-ignore
        this.deviceRendererPlugin = context.deviceRendererPlugin;
    }
    WebGLContextService.prototype.init = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _a, canvas, devicePixelRatio, dpr;
            return __generator(this, function (_b) {
                _a = this.canvasConfig, canvas = _a.canvas, devicePixelRatio = _a.devicePixelRatio;
                this.$canvas = canvas;
                // 实际获取到小程序环境的上下文
                this.context = this.$canvas.getContext('webgl');
                dpr = devicePixelRatio || 1;
                dpr = dpr >= 1 ? Math.ceil(dpr) : 1;
                this.dpr = dpr;
                this.resize(this.canvasConfig.width, this.canvasConfig.height);
                return [2 /*return*/];
            });
        });
    };
    WebGLContextService.prototype.getContext = function () {
        return this.context;
    };
    WebGLContextService.prototype.getDomElement = function () {
        return this.$canvas;
    };
    WebGLContextService.prototype.getDPR = function () {
        return this.dpr;
    };
    WebGLContextService.prototype.getBoundingClientRect = function () {
        if (this.$canvas.getBoundingClientRect) {
            return this.$canvas.getBoundingClientRect();
        }
    };
    WebGLContextService.prototype.destroy = function () {
        // TODO: 小程序环境销毁 context
        this.context = null;
        this.$canvas = null;
    };
    WebGLContextService.prototype.resize = function (width, height) {
        var pixelRatio = devicePixelRatio;
        var canvasDOM = this.$canvas; // HTMLCanvasElement or canvasElement
        // 浏览器环境设置style样式
        // @ts-ignore 使用style功能判断不使用类型判断更准确
        if (canvasDOM.style) {
            // @ts-ignore
            canvasDOM.style.width = width + 'px';
            // @ts-ignore
            canvasDOM.style.height = height + 'px';
        }
        if (isCanvasElement(canvasDOM)) {
            canvasDOM.width = width * pixelRatio;
            canvasDOM.height = height * pixelRatio;
            // if (pixelRatio !== 1) {
            //   this.context.scale(pixelRatio, pixelRatio);
            // }
        }
    };
    WebGLContextService.prototype.applyCursorStyle = function (cursor) {
        // 小程序环境无需设置鼠标样式
    };
    WebGLContextService.prototype.toDataURL = function (options) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.deviceRendererPlugin.toDataURL(options)];
            });
        });
    };
    return WebGLContextService;
}());

var ContextRegisterPlugin = /** @class */ (function (_super) {
    __extends(ContextRegisterPlugin, _super);
    function ContextRegisterPlugin(rendererPlugin) {
        var _this = _super.call(this) || this;
        _this.rendererPlugin = rendererPlugin;
        _this.name = 'mobile-webgl-context-register';
        return _this;
    }
    ContextRegisterPlugin.prototype.init = function () {
        this.context.ContextService = WebGLContextService;
        // @ts-ignore
        this.context.deviceRendererPlugin = this.rendererPlugin;
    };
    ContextRegisterPlugin.prototype.destroy = function () {
        delete this.context.ContextService;
    };
    return ContextRegisterPlugin;
}(AbstractRendererPlugin));

var Renderer = /** @class */ (function (_super) {
    __extends(Renderer, _super);
    function Renderer(config) {
        var _this = _super.call(this, config) || this;
        var deviceRendererPlugin = new DeviceRenderer.Plugin();
        _this.registerPlugin(new ContextRegisterPlugin(deviceRendererPlugin));
        _this.registerPlugin(new ImageLoader.Plugin());
        _this.registerPlugin(new WebGLDevice.Plugin((config === null || config === void 0 ? void 0 : config.targets)
            ? {
                targets: config.targets,
            }
            : {
                targets: ['webgl2', 'webgl1'],
            }));
        _this.registerPlugin(deviceRendererPlugin);
        _this.registerPlugin(new DomInteraction.Plugin());
        _this.registerPlugin(new HTMLRenderer.Plugin());
        _this.registerPlugin(new DragDropEvent.Plugin({
            isDocumentDraggable: isNil(config === null || config === void 0 ? void 0 : config.isDocumentDraggable) ? true : config.isDocumentDraggable,
            isDocumentDroppable: isNil(config === null || config === void 0 ? void 0 : config.isDocumentDroppable) ? true : config.isDocumentDroppable,
            dragstartDistanceThreshold: isNil(config === null || config === void 0 ? void 0 : config.dragstartDistanceThreshold)
                ? 10
                : config.dragstartDistanceThreshold,
            dragstartTimeThreshold: isNil(config === null || config === void 0 ? void 0 : config.dragstartTimeThreshold)
                ? 50
                : config.dragstartTimeThreshold,
        }));
        return _this;
    }
    return Renderer;
}(AbstractRenderer));

export { Renderer };
