// @antv/f-engine/lib/gesture.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _gGesture = _interopRequireDefault(require("@antv/g-gesture"));
var G_SHAPE_EVENT = {
  click: true,
  touchstart: true,
  touchmove: true,
  touchend: true,
  touchendoutside: true,
  dragenter: true,
  dragleave: true,
  dragover: true,
  drop: true,
  dragstart: true,
  drag: true,
  dragend: true
};
var GESTURE_EVENT = {
  panstart: true,
  pan: true,
  panend: true,
  pressstart: true,
  press: true,
  pressend: true,
  swipe: true,
  pinchstart: true,
  pinch: true,
  pinchend: true
};
var Gesture = /** @class */function () {
  function Gesture(element) {
    this.el = element;
    this.gesture = new _gGesture.default(element);
  }
  Gesture.prototype.on = function (eventName, listener) {
    var _a = this,
      el = _a.el,
      gesture = _a.gesture;
    if (G_SHAPE_EVENT[eventName]) {
      el.addEventListener(eventName, listener);
      return;
    }
    if (GESTURE_EVENT[eventName]) {
      gesture.on(eventName, listener);
    }
  };
  Gesture.prototype.off = function (eventName, listener) {
    var _a = this,
      el = _a.el,
      gesture = _a.gesture;
    if (G_SHAPE_EVENT[eventName]) {
      el.removeEventListener(eventName, listener);
      return;
    }
    if (GESTURE_EVENT[eventName]) {
      gesture.off(eventName, listener);
    }
  };
  return Gesture;
}();
var _default = Gesture;
exports.default = _default;
