// @antv/f-engine/lib/timeline.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _tslib = require("tslib");
var _component = _interopRequireDefault(require("./component"));
var _children = _interopRequireDefault(require("./children"));
var Timeline = /** @class */function (_super) {
  (0, _tslib.__extends)(Timeline, _super);
  function Timeline(props) {
    var _this = _super.call(this, props) || this;
    _this.next = function () {
      var _a = _this,
        state = _a.state,
        props = _a.props;
      var index = state.index,
        count = state.count,
        delay = state.delay;
      var loop = props.loop,
        _b = props.autoPlay,
        autoPlay = _b === void 0 ? true : _b;
      var next = loop ? (index + 1) % count : index + 1;
      if (next < count && autoPlay) {
        setTimeout(function () {
          _this.setState({
            index: next
          });
        }, delay || 0);
      }
    };
    var delay = props.delay,
      _a = props.start,
      start = _a === void 0 ? 0 : _a,
      children = props.children;
    var count = _children.default.toArray(children).length;
    _this.state = {
      delay: delay,
      count: count,
      index: start
    };
    return _this;
  }
  Timeline.prototype.didMount = function () {
    var _a = this.props.autoPlay,
      autoPlay = _a === void 0 ? true : _a;
    if (autoPlay) {
      this.animator.on('end', this.next);
    }
  };
  Timeline.prototype.didUnmount = function () {
    this.animator.off('end', this.next);
  };
  Timeline.prototype.render = function () {
    var _a = this,
      state = _a.state,
      props = _a.props;
    var children = props.children;
    var index = state.index;
    var childrenArray = _children.default.toArray(children);
    return childrenArray[index];
  };
  return Timeline;
}(_component.default);
var _default = Timeline;
exports.default = _default;
